bson/src/main/org/bson/codecs/DocumentCodec.java

```diff
@@ -216,10 +229,29 @@ private Object readValue(final BsonReader reader, final DecoderContext decoderCo
             return null;
         } else if (bsonType == BsonType.ARRAY) {
             return readList(reader, decoderContext);
-        } else if (bsonType == BsonType.BINARY && BsonBinarySubType.isUuid(reader.peekBinarySubType()) && reader.peekBinarySize() == 16) {
-            return registry.get(UUID.class).decode(reader, decoderContext);
+        } else {
+            Codec<?> codec = bsonTypeCodecMap.get(bsonType);
+            
+            if (bsonType == BsonType.BINARY && reader.peekBinarySize() == 16) {
+                switch (reader.peekBinarySubType()) {
+                    case 3:
+                        if (uuidRepresentation == UuidRepresentation.JAVA_LEGACY
+                                || uuidRepresentation == UuidRepresentation.C_SHARP_LEGACY
+                                || uuidRepresentation == UuidRepresentation.PYTHON_LEGACY) {
```

Comment:
To preserve the current behavior don't you also need to decode subtype 3 to UUID when the representation is STANDARD? (Basically any representation besides UNSPECIFIED.)

Reply:
There is no current behavior when the representation is STANDARD, because there is currently no way to change the representation in a way that affects the behavior of this class. The only way to mess with it is to register your own Codec<UUID>. If you're using the one that we provide, UuidCodec, then you have to choose the representation.  If you construct it with STANDARD, then if you try to decode a subtype 3 UUID it will throw an exception.

Reply:
Thanks for explaining that.